### PR TITLE
feature: Wan2.2 14B support

### PR DESCRIPTION
**Pull request by kohya-ss**
- [x] update model class
- [x] update inference script
- [x] update training script
- [x] update cache script (CLIP is not needed for Wan2.2)
- [x] update documentation

### CODE REVIEW

**Comment by maybleMyers** [comment]
I got all 3 models working with this https://github.com/maybleMyers/H1111/blob/wan2.2b/wan2_generate_video.py for inference trying to stay close/ use parts of the source repo, it may help if you get stuck. I know your implementation will be better though and am excited for it :D

**Comment by kohya-ss** [comment]
> I got all 3 models working with this https://github.com/maybleMyers/H1111/blob/wan2.2b/wan2_generate_video.py for inference trying to stay close/ use parts of the source repo, it may help if you get stuck. I know your implementation will be better though and am excited for it :D

Thank you! This helps me a lot.

The dynamic loading is very interesting. In my implementation, I decided to load two models in advance to support batch and interactive inference...

**Comment by kohya-ss** [comment]
`--dit_high_noise`, `--timestep_boundary`, `--guidance_scale_high_noise` and `--offload_inactive_dit` options are added to `wan_generate_video.py`

**Comment by maybleMyers** [comment]
The 14B-i2v model seems to work dramatically better with fp32 weights in my tests. With bf16 beating fp16. I did analyze all weights and it seems to fit bf16 with less loss. I am testing it now with a fp32/fp16 mixture.

I have uploaded the bf16 converted i2v model here: https://huggingface.co/maybleMyers/wan_files_for_h1111/tree/main

**Comment by kohya-ss** [comment]
`wan_train_network.py` now supports the 14B model of Wan2.2. The task names are `t2v-A14B` and `i2v-A14B`. There may still be bugs.

Simultaneous training of high and low noise models is not possible. Only one model will be trained.

Please specify the range of time steps to train using the `--min_timestep` and `--max_timestep` options. The newly added `--preserve_distribution_shape` is also recommended. See the document for details: https://github.com/kohya-ss/musubi-tuner/blob/feature-wan-2-2/docs/advanced_config.md#specify-time-step-range-for-training--%E5%AD%A6%E7%BF%92%E6%99%82%E3%81%AE%E3%82%BF%E3%82%A4%E3%83%A0%E3%82%B9%E3%83%86%E3%83%83%E3%83%97%E7%AF%84%E5%9B%B2%E3%81%AE%E6%8C%87%E5%AE%9A

According to the official implementation, the boundary of the timesteps for T2V is 0.875, so when training the low model, min=0, max=875, and for the high model, min=875, max=1000 would be good. Since I2V is 0.9, the thresholds are 900 for both.

**Comment by dorpxam** [comment]
@kohya-ss, did you plan to support the 5B model training in the future? Thank you very much for all, you rock!

**Comment by kohya-ss** [comment]
> @kohya-ss, did you plan to support the 5B model training in the future? Thank you very much for all, you rock!

I plan to support the 5B model as well. However, unlike the 14B model, the 5B model uses the Wan2.2 VAE. Since the Wan2.2 VAE has a different structure from the Wan2.1 VAE, we will need to implement it in this repo. This will likely take some time.

**Comment by dorpxam** [comment]
> I plan to support the 5B model as well. However, unlike the 14B model, the 5B model uses the Wan2.2 VAE. Since the Wan2.2 VAE has a different structure from the Wan2.1 VAE, we will need to implement it in this repo. This will likely take some time.

I perfectly understand. Very cool!

**Comment by kohya-ss** [comment]
Please add `--i2v` for `wan_cache_latents.py` if CLIP is omitted for Wan2.2 I2V training.

**Comment by kohya-ss** [comment]
Support has been added for LoRA training, which applies both high and low models simultaneously. This is highly experimental feature. Specify `--dit_high_noise` and `--timestep_boundary` same as `wan_generate_video.py`.

Sample generation is currently done with a low noise model.

### Without block swap
Even with fp8_scaled, 40GB of VRAM is required for static image or short video training. Specifying `--offload_inactive_dit` halves the required VRAM, but model switching takes longer.

### With block swap
Since both models will be loaded, increase the block swap value. `--offload_inactive_dit` cannot be specified.

File: src/musubi_tuner/wan_train_network.py [hunk 1]
```diff
@@ -191,8 +214,14 @@ def do_inference(
         """architecture dependent inference"""
         model: WanModel = transformer
         device = accelerator.device
+
+        if self.high_low_training:
+            self.next_model_is_high_noise = False # We use low noise model to sample the video
```
[the remark this answers is not in the document]
Comment (kohya-ss) [reply src/musubi_tuner/wan_train_network.py:217]: This will be fixed later.

**Comment by dorpxam** [comment]
> Even with fp8_scaled, 40GB of VRAM is required for static image or short video training. Specifying `--offload_inactive_dit` halves the required VRAM, but model switching takes longer.

Did you mean that 16GB VRAM is not enough for the Wan2.2 14B training, even in very slow training conditions?

**Comment by FurkanGozukara** [comment]
@kohya-ss simultaneous training brings what advantages do you think?

**Comment by kohya-ss** [comment]
> > Even with fp8_scaled, 40GB of VRAM is required for static image or short video training. Specifying `--offload_inactive_dit` halves the required VRAM, but model switching takes longer.
> 
> Did you mean that 16GB VRAM is not enough for the Wan2.2 14B training, even in very slow training conditions?

Training LoRA with a single model works fine with 16GB VRAM with block swap. I'm not sure training with both models with a larger block swap value would probably work, but I'm not sure.

**Comment by kohya-ss** [comment]
> @kohya-ss simultaneous training brings what advantages do you think?

It's still unclear if it will work or not, but there are reports that the high noise model handles character styles and features, so it may be possible to improve the quality.

**Comment by dorpxam** [comment]
> Training LoRA with a single model works fine with 16GB VRAM with block swap. I'm not sure training with both models with a larger block swap value would probably work, but I'm not sure.

Oh cool! I will try that soon. I've try the 5B model in inference mode but I'm not sure that the result will be very good. The I2V work approximately well except in the case of characters who are relatively distant. Squinting eyes or other defects in the details. The T2V works less well. Many problems in large movements, other than subtle movements in more or less close-up, something close to the LTXVideo. This is why despite the fact that the 5B model runs at 24 images per second, I prefer to fall back on the 14B which in version 2.1 already gave me satisfaction. This, I will not fail to test the training of the 5B model, who knows. Maybe the LoRA can generate surprising results.

So very happy to hear that 16GB can fit for training! Thank's man!

**Comment by sdbds** [comment]
Is there any configuration for video datasets? It's strange that my test shows significantly higher GPU memory usage than the old version... (480P using `frame_extraction=full` max_frames=81)

**Comment by DarkJukeMastah** [comment]
I would also be interested in a working config for video training.
Using a 4090, both the VRAM and shared VRAM fill completely with 20 swapped blocks and 1 worker and the training basically dies.

**Comment by kohya-ss** [comment]
I will investigate memory usage issues.

There seems to be an issue with transformers 4.54.1 where GPU memory is not released even when the Text Encoder is deleted. If you have upgraded your transformers version, try reverting to 4.46.3.

**Comment by sdbds** [comment]
> I will investigate memory usage issues.
> 
> There seems to be an issue with transformers 4.54.1 where GPU memory is not released even when the Text Encoder is deleted. If you have upgraded your transformers version, try reverting to 4.46.3.

It seems so... So that's it...Always issues from huggingface dependencies...

**Comment by DarkJukeMastah** [comment]
> I will investigate memory usage issues.
> 
> There seems to be an issue with transformers 4.54.1 where GPU memory is not released even when the Text Encoder is deleted. If you have upgraded your transformers version, try reverting to 4.46.3.

I tried to downgrade my transformers version, and it turns out I was already on 4.46.3.

For reference, this is what my usage looks like after a few minutes. My shared memory (bottom) just keeps climbing until it fully fills up.
image

This is what I'm trying to run:
accelerate launch --num_cpu_threads_per_process 1 src/musubi_tuner/wan_train_network.py --task t2v-14B --dit "C:/Python/ComfyUI/models/diffusion_models/wan2.2_t2v_high_noise_14B_fp16.safetensors" --t5 "C:/Python/ComfyUI/models/text_encoders/umt5-xxl-enc-fp8_e4m3fn.safetensors" --dataset_config "C:/Python/vid/dataset.toml" --sdpa --mixed_precision fp16 --fp8_base --fp8_scaled --vae "C:/Python/ComfyUI/models/vae/wan_2.1_vae.safetensors" --blocks_to_swap 20 --optimizer_type adamw8bit --learning_rate 2e-4 --gradient_checkpointing --max_data_loader_n_workers 1 --persistent_data_loader_workers --network_module networks.lora_wan --network_dim 32 --timestep_sampling shift --discrete_flow_shift 3.0 --max_train_epochs 16 --save_every_n_epochs 1 --seed 42 --output_dir "C:/Python/vid/output/"

**Comment by hablaba** [comment]
With these new timestep args, how do those play with shift X vs sigmoid vs uniform? Like normally shift can push training to low or high time steps and sigmoid trains kind of in the middle. But when we specify the range at say the high time steps, does the distribution of training match to that range? Some graphs may be helpful in docs. 

For training on Wan 2.1 I preferred sigmoid. In wan 2.2 it seems to work ok but it’s less clear to me what it’s doing with the timestep configuration included.

**Comment by Tophness** [comment]
> Simultaneous training of high and low noise models is not possible. Only one model will be trained.

When I provide the high noise model to --dit_high_noise I get this error asking for the low noise model. Why does it need the other model if only 1 model is being trained?

```
  File "E:\Downloads\musubi-tuner\src\musubi_tuner\hv_train_network.py", line 1455, in train
    raise ValueError("path to DiT model is required / DiTモデルのパスが必要です")
```

**Comment by kohya-ss** [comment]
> When I provide the high noise model to --dit_high_noise I get this error asking for the low noise model. Why does it need the other model if only 1 model is being trained?

When training a single model, even if you are training a high model, you should specify the `--dit` option.

**Comment by kohya-ss** [comment]
I merged #414 to revert the dtype of the loaded model. I hope this fixes some memory issues.

> For reference, this is what my usage looks like after a few minutes. My shared memory (bottom) just keeps climbing until it fully fills up.

The options seem to be fine. What is the image size and number of frames in your dataset? Long, high-resolution videos consume a lot of memory.

**Comment by kohya-ss** [comment]
> For training on Wan 2.1 I preferred sigmoid. In wan 2.2 it seems to work ok but it’s less clear to me what it’s doing with the timestep configuration included.

You can visualize the distribution shape of the timesteps with `--show_timesteps image` (or `--show_timesteps console`) option. Try changing the options related to timesteps to see the actual distribution.

**Comment by DarkJukeMastah** [comment]
> I merged #414 to revert the dtype of the loaded model. I hope this fixes some memory issues.
> 
> > For reference, this is what my usage looks like after a few minutes. My shared memory (bottom) just keeps climbing until it fully fills up.
> 
> The options seem to be fine. What is the image size and number of frames in your dataset? Long, high-resolution videos consume a lot of memory.

I'm just trying to test stuff, so I'm only trying with five 8 second 720p videos.

This is what I'm using in the dataset.toml file:
[general]
resolution = [640, 480]
caption_extension = ".txt"
batch_size = 1
enable_bucket = true
bucket_no_upscale = false

[[datasets]]
video_directory = "C:/Python/vid"
cache_directory = "C:/Python/vid/cache" # recommended to set cache directory
frame_extraction = "full"
source_fps = 16.0 # optional, source fps for videos in the directory, decimal number

**Comment by sdbds** [comment]
> > I merged #414 to revert the dtype of the loaded model. I hope this fixes some memory issues.
> > > For reference, this is what my usage looks like after a few minutes. My shared memory (bottom) just keeps climbing until it fully fills up.
> > 
> > 
> > The options seem to be fine. What is the image size and number of frames in your dataset? Long, high-resolution videos consume a lot of memory.
> 
> I'm just trying to test stuff, so I'm only trying with five 8 second 720p videos.
> 
> This is what I'm using in the dataset.toml file: [general] resolution = [640, 480] caption_extension = ".txt" batch_size = 1 enable_bucket = true bucket_no_upscale = false
> 
> [[datasets]] video_directory = "C:/Python/vid" cache_directory = "C:/Python/vid/cache" # recommended to set cache directory frame_extraction = "full" source_fps = 16.0 # optional, source fps for videos in the directory, decimal number

The main issue comes from the sampling frame extraction method; changing it to uniform will significantly reduce GPU memory usage.

**Comment by DarkJukeMastah** [comment]
> > > I merged #414 to revert the dtype of the loaded model. I hope this fixes some memory issues.
> > > > For reference, this is what my usage looks like after a few minutes. My shared memory (bottom) just keeps climbing until it fully fills up.
> > > 
> > > 
> > > The options seem to be fine. What is the image size and number of frames in your dataset? Long, high-resolution videos consume a lot of memory.
> > 
> > 
> > I'm just trying to test stuff, so I'm only trying with five 8 second 720p videos.
> > This is what I'm using in the dataset.toml file: [general] resolution = [640, 480] caption_extension = ".txt" batch_size = 1 enable_bucket = true bucket_no_upscale = false
> > [[datasets]] video_directory = "C:/Python/vid" cache_directory = "C:/Python/vid/cache" # recommended to set cache directory frame_extraction = "full" source_fps = 16.0 # optional, source fps for videos in the directory, decimal number
> 
> The main issue comes from the sampling frame extraction method; changing it to uniform will significantly reduce GPU memory usage.

Damn, that was the problem!
Thank you, sorry for spamming the thread.

**Comment by kohya-ss** [comment]
I've merged this PR. Thanks to everyone. 

I'll be working on the 5B model in a separate PR, but I might prioritize Qwen-Image.

**Comment by FurkanGozukara** [comment]
> I've merged this PR. Thanks to everyone.
> 
> I'll be working on the 5B model in a separate PR, but I might prioritize Qwen-Image.

qwen image please ty

**Comment by sdbds** [comment]
https://www.reddit.com/r/StableDiffusion/comments/1mkv9c6/wan22_schedulers_steps_shift_and_noise/
image
Just saw the discussion on Reddit today, and I thought perhaps T2V/I2V models need to use different shifts...

**Comment by Lumoria** [comment]
When training a lora for both high and low models at the same time, only one lora model is being saved. I assume we just load the same lora on both the high and low models when we run inference? Or am I misunderstanding this and we should be training the high and low loras in two separate runs?

**Comment by kohya-ss** [comment]
> Just saw the discussion on Reddit today, and I thought perhaps T2V/I2V models need to use different shifts...

Thank you, I've updated the documentation: #422.

**Comment by kohya-ss** [comment]
> When training a lora for both high and low models at the same time, only one lora model is being saved. I assume we just load the same lora on both the high and low models when we run inference? Or am I misunderstanding this and we should be training the high and low loras in two separate runs?

Yes, just load the same LoRA model for both high noise and low noise.

**Comment by Tophness** [comment]
> Just saw the discussion on Reddit today, and I thought perhaps T2V/I2V models need to use different shifts...

Anyone tested the difference? I've already started a long I2V training run with a 3.0 shift. Is it worth starting again with a 5.0?
And what's the relation to training shift during inference? If it was trained on 3.0 do I need to use 3.0 for inference?


> When training a lora for both high and low models at the same time

Is it safe to say we can also train low and high models with completely separate runs and then do the switching between them during inference?
Sorry I know these are nooby questions, but there's not much clear advise elsewhere

**Comment by WepeNerd** [comment]
> The 14B-i2v model seems to work dramatically better with fp32 weights in my tests. With bf16 beating fp16. I did analyze all weights and it seems to fit bf16 with less loss. I am testing it now with a fp32/fp16 mixture.
> 
> I have uploaded the bf16 converted i2v model here: https://huggingface.co/maybleMyers/wan_files_for_h1111/tree/main

Have you come to a conclusion on which model is best?
